cmake/test/ssp/test.cpp

```diff
@@ -0,0 +1,12 @@
+#include <cstring>
+
+void f(const char* s)
```

Comment:
Can you elaborate on in which scenario you tested, that this fails? The code here compiles fine with `x86_64-w64-mingw32-g++ (GCC) 10-posix 20220113` and `mingw-w64 version 8` (because GCC automatically adds `-lssp` when using `-fstack-protector-(strong|all)`), nothing todo with Autotools/libtool, or the mingw-w64 version etc.

i.e:
```bash
cat test.cpp 
#include <cstring>

void f(const char* s)
{
    char b[42];
    std::strcpy(b, s);
}

int main()
{
    f("test");
}
```
```bash
x86_64-w64-mingw32-g++ --version
# x86_64-w64-mingw32-g++ (GCC) 10-posix 20220113

x86_64-w64-mingw32-g++ -E -dM test.cpp | grep __MINGW64_VERSION_MAJOR
#define __MINGW64_VERSION_MAJOR 8
```

```bash
x86_64-w64-mingw32-g++ test.cpp -v -fstack-protector-all -v
Using built-in specs.
COLLECT_GCC=x86_64-w64-mingw32-g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-w64-mingw32/10-posix/lto-wrapper
Target: x86_64-w64-mingw32
Configured with: ../../src/configure --build=aarch64-linux-gnu --prefix=/usr --includedir='/usr/include' --mandir='/usr/share/man' --infodir='/usr/share/info' --sysconfdir=/etc --localstatedir=/var --disable-option-checking --disable-silent-rules --libdir='/usr/lib/aarch64-linux-gnu' --libexecdir='/usr/lib/aarch64-linux-gnu' --disable-maintainer-mode --disable-dependency-tracking --prefix=/usr --enable-shared --enable-static --disable-multilib --with-system-zlib --libexecdir=/usr/lib --without-included-gettext --libdir=/usr/lib --enable-libstdcxx-time=yes --with-tune=generic --with-headers --enable-version-specific-runtime-libs --enable-fully-dynamic-string --enable-libgomp --enable-languages=c,c++,fortran,objc,obj-c++,ada --enable-lto --enable-threads=posix --program-suffix=-posix --program-prefix=x86_64-w64-mingw32- --target=x86_64-w64-mingw32 --with-as=/usr/bin/x86_64-w64-mingw32-as --with-ld=/usr/bin/x86_64-w64-mingw32-ld --enable-libatomic --enable-libstdcxx-filesystem-ts=yes --enable-dependency-tracking SED=/bin/sed
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 10-posix 20220113 (GCC) 
COLLECT_GCC_OPTIONS='-v' '-fstack-protector-all' '-v' '-shared-libgcc' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/cc1plus -quiet -v -v -D_REENTRANT test.cpp -quiet -dumpbase test.cpp -mtune=generic -march=x86-64 -auxbase test -version -fstack-protector-all -o /tmp/ccqYCmIW.s
GNU C++14 (GCC) version 10-posix 20220113 (x86_64-w64-mingw32)
	compiled by GNU C version 10.3.0, GMP version 6.2.1, MPFR version 4.1.0, MPC version 1.2.1, isl version isl-0.24-GMP

GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
ignoring nonexistent directory "/usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/sys-include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/include/c++
 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/include/c++/x86_64-w64-mingw32
 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/include/c++/backward
 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/include
 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/include-fixed
 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/include
End of search list.
GNU C++14 (GCC) version 10-posix 20220113 (x86_64-w64-mingw32)
	compiled by GNU C version 10.3.0, GMP version 6.2.1, MPFR version 4.1.0, MPC version 1.2.1, isl version isl-0.24-GMP

GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
Compiler executable checksum: 316e34e5e802698d4a8bd90af2851cb1
COLLECT_GCC_OPTIONS='-v' '-fstack-protector-all' '-v' '-shared-libgcc' '-mtune=generic' '-march=x86-64'
 /usr/bin/x86_64-w64-mingw32-as -v -v -o /tmp/ccRUOIuh.o /tmp/ccqYCmIW.s
GNU assembler version 2.38 (x86_64-w64-mingw32) using BFD version (GNU Binutils) 2.38
COMPILER_PATH=/usr/lib/gcc/x86_64-w64-mingw32/10-posix/:/usr/lib/gcc/x86_64-w64-mingw32/10-posix/:/usr/lib/gcc/x86_64-w64-mingw32/:/usr/lib/gcc/x86_64-w64-mingw32/10-posix/:/usr/lib/gcc/x86_64-w64-mingw32/:/usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/bin/
LIBRARY_PATH=/usr/lib/gcc/x86_64-w64-mingw32/10-posix/:/usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/lib/
COLLECT_GCC_OPTIONS='-v' '-fstack-protector-all' '-v' '-shared-libgcc' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-w64-mingw32/10-posix/collect2 -plugin /usr/lib/gcc/x86_64-w64-mingw32/10-posix/liblto_plugin.so -plugin-opt=/usr/lib/gcc/x86_64-w64-mingw32/10-posix/lto-wrapper -plugin-opt=-fresolution=/tmp/ccdYkWdb.res -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -plugin-opt=-pass-through=-lpthread -plugin-opt=-pass-through=-ladvapi32 -plugin-opt=-pass-through=-lshell32 -plugin-opt=-pass-through=-luser32 -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -m i386pep -Bdynamic /usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/lib/crt2.o /usr/lib/gcc/x86_64-w64-mingw32/10-posix/crtbegin.o -L/usr/lib/gcc/x86_64-w64-mingw32/10-posix -L/usr/lib/gcc/x86_64-w64-mingw32/10-posix/../../../../x86_64-w64-mingw32/lib /tmp/ccRUOIuh.o -lstdc++ -lssp_nonshared -lssp -lmingw32 -lgcc_s -lgcc -lmoldname -lmingwex -lmsvcrt -lpthread -ladvapi32 -lshell32 -luser32 -lkernel32 -lmingw32 -lgcc_s -lgcc -lmoldname -lmingwex -lmsvcrt /usr/lib/gcc/x86_64-w64-mingw32/10-posix/crtend.o
COLLECT_GCC_OPTIONS='-v' '-fstack-protector-all' '-v' '-shared-libgcc' '-mtune=generic' '-march=x86-64'
```

Reply:
> Can you elaborate on in which scenario you tested, that this fails?

Separate compiling and linking stages?

```
$ x86_64-w64-mingw32-g++ test.cpp -fstack-protector-all -o test.o
$ x86_64-w64-mingw32-g++ test.o -o test.exe

Reply:
> Can you elaborate on in which scenario you tested, that this fails?

Separate compiling and linking stages?

```
$ x86_64-w64-mingw32-g++ -c -fstack-protector-all test.cpp -o test.o
$ x86_64-w64-mingw32-g++ test.o -o test.exe
/usr/bin/x86_64-w64-mingw32-ld: test.o:test.cpp:(.text+0x3e): undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: test.o:test.cpp:(.text+0x8a): undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: test.o:test.cpp:(.rdata$.refptr.__stack_chk_guard[.refptr.__stack_chk_guard]+0x0): undefined reference to `__stack_chk_guard'
collect2: error: ld returned 1 exit status
```

Reply:
Right, I guess that isn't really representative of of how we compile things, so is it the best test? If GCC is always going to be adding `-lssp` in any case, when we pass `-fstack-protector-*`, it's still not clear to me what we are trying to "fix" or work around here.

Reply:
> nothing todo with Autotools/libtool

You are right. A reference to Autotools in the comment is wrong.

Reply:
The CMake-specific aspect is that the [`target_compile_options`](https://cmake.org/cmake/help/latest/command/target_compile_options.html) command affects only a compiler context, while the [`target_link_options`](https://cmake.org/cmake/help/latest/command/target_link_options.html) command operates exclusively within a linker context.

Using a compiler driver with clearly separated compiling and linking contexts is not a CMake-specific issue because the Autotools do the same, no?

Reply:
> If GCC is always going to be adding `-lssp` in any case, when we pass `-fstack-protector-*`, it's still not clear to me what we are trying to "fix" or work around here.

Is it true when `-fstack-protector-*` is passed only to a compiling context, not to a linking one?

Reply:
I think I'm caught up with what's going on here. Apologies if I'm duplicating some of the above, but here are my findings:

An ssp impl was added for mingw-w64 11. You can see the history here: https://github.com/mingw-w64/mingw-w64/commits/master/mingw-w64-crt/ssp
This allows libssp to be an empty dummy file, meaning that `-lssp` will still work. **Given how new this is, I don't think it's really relevant to us yet.**

gcc will pass -lssp automatically during the linking stage, but not when compiling. Consider this broken sequence:
```
$ x86_64-w64-mingw32-g++ -O2 test.cpp -fstack-protector-all -c -o test.o
$ x86_64-w64-mingw32-g++ test.o -o testing
/usr/bin/x86_64-w64-mingw32-ld: test.o:test.cpp:(.text.startup+0x2c): undefined reference to `__stack_chk_fail'
```
It fails because the compile with `-fstack-protector-all` caused an unresolved`__stack_chk_fail` to be emitted, but then nothing resolves it at link-time.

This one, however, works:
```
$ x86_64-w64-mingw32-g++ -O2 test.cpp -fstack-protector-all -c -o test.o
$ x86_64-w64-mingw32-g++ test.o -fstack-protector-all -o testing
```

As does this one:
```
$ x86_64-w64-mingw32-g++ -O2 test.cpp -fstack-protector-all -o testing
```

(Adding -v to get a verbose link-line shows `-lssp_nonshared -lssp` being added by `-fstack-protector-all`)

If I'm understanding correctly, the difference in behavior we're seeing with master is that our CXXFLAGS don't make it to the link line. So, I think the solution is as simple as adding `-fstack-protector-all` to LDFLAGS as well as CXXFLAGS?

Reply:
> So, I think the solution is as simple as adding `-fstack-protector-all` to LDFLAGS as well as CXXFLAGS?

Thanks! Done.

Reply:
I'm pretty sure we've been here before, and I'm pretty sure my pushback was against adding what is a compilation  flag, to link flags, just to invoke a side-effect. Can we at least document it, because I don't think it's going to be immediately obvious why this is here/needed.

Reply:
> I'm pretty sure we've been here before, and I'm pretty sure my pushback was against adding what is a compilation flag, to link flags, just to invoke a side-effect.

You are right.

> Can we at least document it, because I don't think it's going to be immediately obvious why this is here/needed.

Mind suggesting a nice-worded comment please?

Reply:
I guess it's fair to say this being used to "invoke a side-effect", but I think it's reasonable to let the compiler driver do the right thing here as opposed to us figuring out. I suspect you don't like "this is how it currently works anyway" as a justification, but that's true too :)

If anything, I expected you to grump at the idea that we're no longer forwarding all CXXFLAGS to LDFLAGS. I find that to be a nice change as it allows us to be more explicit, but I wouldn't be surprised if there are other flags that we need to copy to LDFLAGS now (maybe sanitizers? lto? etc) as well.

Reply:
>... adding what is a compilation flag, to link flags...

According to the GCC [docs](https://gcc.gnu.org/onlinedocs/gcc/Invoking-GCC.html):
> When you invoke GCC, it normally does preprocessing, compilation, assembly and linking. ...options are passed on to one or more stages of processing.

Moreover, there are [other options](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#index-flto) that the documentation explicitly mentions:
> It is recommended that you compile all the files participating in the same link with the same options and also specify those options at link time.

So, my analysis concludes that the notions of "a compilation flag" and "a link flag" are confusing. Instead, there are flags/options that are passed on to compilation only, or compilation and linking, or linking only, or some other combination of the four stages.

As the `-fstack-protector-all` option might affect the linking stage, it should be passed on to the linking stage. It is not " just to invoke a side-effect" in a negative connotation. It is aligned with the behavior of other options by design, however, not clearly documented.

---

CMake [refers to](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS.html) such options as "language-wide flags":
> These flags will be passed to all invocations of the compiler. This includes invocations that drive compiling and those that drive linking.

On the other hand, the [`target_compile_options`](https://cmake.org/cmake/help/latest/command/target_compile_options.html) and [`target_link_options`](https://cmake.org/cmake/help/latest/command/target_link_options.html) commands are stage-specific, which is nice. Therefore, it is correct and intuitive to use both commands to pass a language-wide option to a target.

Reply:
> Can we at least document it, because I don't think it's going to be immediately obvious why this is here/needed.

Sure. A comment has been added.

Reply:
Unrelated since https://github.com/hebasto/bitcoin/pull/74#issuecomment-1927621595. Resolving.